# Exploration du dataset tips

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns

In [3]:
tips = sns.load_dataset("tips")

## Structure du dataset

In [4]:
tips.shape

(244, 7)

In [5]:
tips.head(10)

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4
5,25.29,4.71,Male,No,Sun,Dinner,4
6,8.77,2.00,Male,No,Sun,Dinner,2
7,26.88,3.12,Male,No,Sun,Dinner,4
8,15.04,1.96,Male,No,Sun,Dinner,2
9,14.78,3.23,Male,No,Sun,Dinner,2


In [6]:
tips.info()

<class 'pandas.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   total_bill  244 non-null    float64 
 1   tip         244 non-null    float64 
 2   sex         244 non-null    category
 3   smoker      244 non-null    category
 4   day         244 non-null    category
 5   time        244 non-null    category
 6   size        244 non-null    int64   
dtypes: category(4), float64(2), int64(1)
memory usage: 7.4 KB


In [7]:
tips.describe()

,total_bill,tip,size
count,244.000000,244.000000,244.000000
mean,19.785943,2.998279,2.569672
std,8.902412,1.383638,0.951100
min,3.070000,1.000000,1.000000
25%,13.347500,2.000000,2.000000
50%,17.795000,2.900000,2.000000
75%,24.127500,3.562500,3.000000
max,50.810000,10.000000,6.000000


## Types de données

In [8]:
tips.dtypes

total_bill     float64
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
dtype: object

## Classification des variables

In [9]:
tips["total_bill"].sample(10).tolist()

[25.21, 21.01, 16.31, 11.02, 16.82, 31.27, 26.86, 22.75, 9.6, 13.0]

In [10]:
tips["total_bill"].value_counts()

total_bill
13.42    3
10.34    2
21.01    2
10.33    2
17.92    2
        ..
29.03    1
27.18    1
22.67    1
17.82    1
18.78    1
Name: count, Length: 229, dtype: int64

- `total_bill` est une variable quantitative continue avec 229 valeurs differentes possibles sur 244 lignes

In [11]:
tips["tip"].value_counts()

tip
2.00    33
3.00    23
4.00    12
5.00    10
2.50    10
        ..
1.47     1
1.17     1
4.67     1
5.92     1
1.75     1
Name: count, Length: 123, dtype: int64

- `tip` est une variable quantitative continue avec 123 valeurs differentes possibles sur 244 lignes

In [12]:
tips["size"].value_counts()

size
2    156
3     38
4     37
5      5
1      4
6      4
Name: count, dtype: int64

In [13]:
tips["sex"].value_counts()

sex
Male      157
Female     87
Name: count, dtype: int64

In [14]:
tips["day"].value_counts()

day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64

In [15]:
tips["time"].value_counts()

time
Dinner    176
Lunch      68
Name: count, dtype: int64

In [16]:
tips["smoker"].value_counts()

smoker
No     151
Yes     93
Name: count, dtype: int64

- `size` est une variable quantitative discrete (discontinu) avec 5 valeurs differentes possibles sur 244 lignes

- `sex`, `smoker`, `date` et `time` sont des variables catégorielles

## Valeurs manquantes 

In [17]:
tips.isna().mean()

total_bill    0.0
tip           0.0
sex           0.0
smoker        0.0
day           0.0
time          0.0
size          0.0
dtype: float64

In [18]:
tips.isna().sum()

total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64

**Constat** : Une moyenne de 0% de valeurs manquantes et une absence de lignes avec des valeurs manquantes dans le dataset

## Doublons

In [19]:
tips[tips.duplicated()]

,total_bill,tip,sex,smoker,day,time,size
202,13.0,2.0,Female,Yes,Thur,Lunch,2


In [20]:
tips[tips.duplicated(keep='last')]

,total_bill,tip,sex,smoker,day,time,size
198,13.0,2.0,Female,Yes,Thur,Lunch,2


In [21]:
tips.value_counts().head()

total_bill  tip   sex     smoker  day   time    size
13.00       2.00  Female  Yes     Thur  Lunch   2       2
16.99       1.01  Female  No      Sun   Dinner  2       1
10.34       1.66  Male    No      Sun   Dinner  3       1
21.01       3.50  Male    No      Sun   Dinner  3       1
23.68       3.31  Male    No      Sun   Dinner  2       1
Name: count, dtype: int64

**Constat** : Présence d'un doublon présent sur la ligne 202

In [22]:
tips[tips["total_bill"].duplicated()]

,total_bill,tip,sex,smoker,day,time,size
60,20.29,3.21,Male,Yes,Sat,Dinner,2
63,18.29,3.76,Male,Yes,Sat,Dinner,4
76,17.92,3.08,Male,Yes,Sat,Dinner,2
98,21.01,3.00,Male,Yes,Fri,Dinner,2
136,10.33,2.00,Female,No,Thur,Lunch,2
163,13.81,2.00,Male,No,Sun,Dinner,2
172,7.25,5.15,Male,Yes,Sun,Dinner,2
185,20.69,5.00,Male,No,Sun,Dinner,5
196,10.34,2.00,Male,Yes,Thur,Lunch,2
202,13.00,2.00,Female,Yes,Thur,Lunch,2


In [23]:
tips["total_bill"].duplicated().sum(), tips["total_bill"].duplicated().mean()

(np.int64(15), np.float64(0.06147540983606557))

In [24]:
tips[tips["tip"].duplicated()]

,total_bill,tip,sex,smoker,day,time,size
18,16.97,3.50,Female,No,Sun,Dinner,3
26,13.37,2.00,Male,No,Sat,Dinner,2
27,12.69,2.00,Male,No,Sat,Dinner,2
29,19.65,3.00,Female,No,Sat,Dinner,2
32,15.06,3.00,Female,No,Sat,Dinner,2
...,...,...,...,...,...,...,...
235,10.07,1.25,Male,No,Sat,Dinner,2
236,12.60,1.00,Male,Yes,Sat,Dinner,2
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2


In [25]:
tips["tip"].duplicated().sum(), tips["tip"].duplicated().mean()

(np.int64(121), np.float64(0.4959016393442623))

In [26]:
tips[tips["size"].duplicated()]

,total_bill,tip,sex,smoker,day,time,size
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
5,25.29,4.71,Male,No,Sun,Dinner,4
6,8.77,2.00,Male,No,Sun,Dinner,2
7,26.88,3.12,Male,No,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2
242,17.82,1.75,Male,No,Sat,Dinner,2


In [27]:
tips["size"].duplicated().sum(), tips["size"].duplicated().mean()

(np.int64(238), np.float64(0.9754098360655737))

## Dates

In [28]:
tips['day'].value_counts()

day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64

In [29]:
tips['day'].dtype

CategoricalDtype(categories=['Thur', 'Fri', 'Sat', 'Sun'], ordered=False, categories_dtype=str)

In [30]:
tips["day"].sample(10).values.tolist()

['Thur', 'Sat', 'Sun', 'Sun', 'Fri', 'Sun', 'Thur', 'Thur', 'Thur', 'Sun']

**Constat**: La varaible `day` est de type `category` qui permet de conserver l'ordre temporel des jours plutot que l'ordre alphabétique des `str`.

## Analyse de corrélation

### Analyse de la corrélation entre le montant de la facture, le pourboire et le sexe des clients

In [32]:
tips[["total_bill", "tip"]].describe()

,total_bill,tip
count,244.000000,244.000000
mean,19.785943,2.998279
std,8.902412,1.383638
min,3.070000,1.000000
25%,13.347500,2.000000
50%,17.795000,2.900000
75%,24.127500,3.562500
max,50.810000,10.000000


In [33]:
tips.loc[:, ["total_bill", "tip"]].groupby(tips["sex"]).sum()

,total_bill,tip
sex,,
Male,3256.82,485.07
Female,1570.95,246.51


In [34]:
tips.loc[:, ["total_bill", "tip"]].groupby(tips["sex"]).count()

,total_bill,tip
sex,,
Male,157,157
Female,87,87


In [35]:
tips.loc[:, ["total_bill", "tip"]].groupby(tips["sex"]).mean()

,total_bill,tip
sex,,
Male,20.744076,3.089618
Female,18.056897,2.833448


### Analyse de la corrélation entre le montant de la facture, le pourboire et une table de fumeurs ou de non fumeurs

In [36]:
tips.loc[:, ["total_bill", "tip"]].groupby(tips["smoker"]).count()

,total_bill,tip
smoker,,
Yes,93,93
No,151,151


In [37]:
tips.loc[:, ["total_bill", "tip"]].groupby(tips["smoker"]).sum()

,total_bill,tip
smoker,,
Yes,1930.34,279.81
No,2897.43,451.77


In [38]:
tips.loc[:, ["total_bill", "tip"]].groupby(tips["smoker"]).mean()

,total_bill,tip
smoker,,
Yes,20.756344,3.008710
No,19.188278,2.991854


### Analyse de la corrélation entre le montant de la facture, le pourboire et le nombre de personnes par table

In [39]:
tips.loc[:, ["total_bill", "tip"]].groupby(tips["size"]).count()

,total_bill,tip
size,,
1,4,4
2,156,156
3,38,38
4,37,37
5,5,5
6,4,4


In [40]:
tips.loc[:, ["total_bill", "tip"]].groupby(tips["size"]).sum()

,total_bill,tip
size,,
1,28.97,5.75
2,2565.89,402.84
3,884.55,128.94
4,1058.70,153.01
5,150.34,20.14
6,139.32,20.90


In [41]:
tips.loc[:, ["total_bill", "tip"]].groupby(tips["size"]).mean()

,total_bill,tip
size,,
1,7.242500,1.437500
2,16.448013,2.582308
3,23.277632,3.393158
4,28.613514,4.135405
5,30.068000,4.028000
6,34.830000,5.225000


### Analyse de la corrélation entre le montant de la facture, le pourboire, le nombre de personnes et le jour du repas

In [42]:
tips.loc[:, ["total_bill", "tip", "size"]].groupby(tips["day"]).count()

,total_bill,tip,size
day,,,
Thur,62,62,62
Fri,19,19,19
Sat,87,87,87
Sun,76,76,76


In [43]:
tips.loc[:, ["total_bill", "tip", "size"]].groupby(tips["day"]).sum()

,total_bill,tip,size
day,,,
Thur,1096.33,171.83,152
Fri,325.88,51.96,40
Sat,1778.40,260.40,219
Sun,1627.16,247.39,216


In [44]:
tips.loc[:, ["total_bill", "tip", "size"]].groupby(tips["day"]).mean()

,total_bill,tip,size
day,,,
Thur,17.682742,2.771452,2.451613
Fri,17.151579,2.734737,2.105263
Sat,20.441379,2.993103,2.517241
Sun,21.410000,3.255132,2.842105


### Analyse de la corrélation entre le montant de la facture, le pourboire, le nombre de personnes et le moment du repas

In [45]:
tips.loc[:, ["total_bill", "tip", "size"]].groupby(tips["time"]).count()

,total_bill,tip,size
time,,,
Lunch,68,68,68
Dinner,176,176,176


In [46]:
tips.loc[:, ["total_bill", "tip", "size"]].groupby(tips["time"]).sum()

,total_bill,tip,size
time,,,
Lunch,1167.47,185.51,164
Dinner,3660.30,546.07,463


In [47]:
tips.loc[:, ["total_bill", "tip", "size"]].groupby(tips["time"]).mean()

,total_bill,tip,size
time,,,
Lunch,17.168676,2.728088,2.411765
Dinner,20.797159,3.102670,2.630682


## Détection d'anomalies

### `total_bill`

In [48]:
tips["total_bill"].head(20).values.tolist()

[16.99,
 10.34,
 21.01,
 23.68,
 24.59,
 25.29,
 8.77,
 26.88,
 15.04,
 14.78,
 10.27,
 35.26,
 15.42,
 18.43,
 14.83,
 21.58,
 10.33,
 16.29,
 16.97,
 20.65]

### `tip`

In [49]:
tips["tip"].head(20).values.tolist()

[1.01,
 1.66,
 3.5,
 3.31,
 3.61,
 4.71,
 2.0,
 3.12,
 1.96,
 3.23,
 1.71,
 5.0,
 1.57,
 3.0,
 3.02,
 3.92,
 1.67,
 3.71,
 3.5,
 3.35]

- **Constat**: La variable est de type float64

- **Action**: fixer la variable à 2 décimales près

### `sex`

In [50]:
tips["sex"].dtype

CategoricalDtype(categories=['Male', 'Female'], ordered=False, categories_dtype=str)

In [51]:
tips["sex"].head(20).values.tolist()

['Female',
 'Male',
 'Male',
 'Male',
 'Female',
 'Male',
 'Male',
 'Male',
 'Male',
 'Male',
 'Male',
 'Female',
 'Male',
 'Male',
 'Female',
 'Male',
 'Female',
 'Male',
 'Female',
 'Male']

In [52]:
tips[(tips["sex"] == "Female") | (tips["sex"] == "Male")].count()

total_bill    244
tip           244
sex           244
smoker        244
day           244
time          244
size          244
dtype: int64

### `smoker`

In [53]:
tips["smoker"].dtype

CategoricalDtype(categories=['Yes', 'No'], ordered=False, categories_dtype=str)

In [54]:
tips["smoker"].head(20).values.tolist()

['No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No',
 'No']

In [55]:
tips[(tips["smoker"] == "No") | (tips["smoker"] == "Yes")].count()

total_bill    244
tip           244
sex           244
smoker        244
day           244
time          244
size          244
dtype: int64

### `day`

In [56]:
tips["day"].dtype

CategoricalDtype(categories=['Thur', 'Fri', 'Sat', 'Sun'], ordered=False, categories_dtype=str)

In [57]:
tips["day"].head(20).values.tolist()

['Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sun',
 'Sat']

In [58]:
tips[(tips["day"] == "Thur") | (tips["day"] == "Fri") | (tips["day"] == "Sat") | (tips["day"] == "Sun")].count()

total_bill    244
tip           244
sex           244
smoker        244
day           244
time          244
size          244
dtype: int64

### `time`

In [59]:
tips["time"].dtype

CategoricalDtype(categories=['Lunch', 'Dinner'], ordered=False, categories_dtype=str)

In [60]:
tips["time"].head(20).values.tolist()

['Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner',
 'Dinner']

In [61]:
tips[(tips["time"] == "Lunch") | (tips["time"] == "Dinner")].count()

total_bill    244
tip           244
sex           244
smoker        244
day           244
time          244
size          244
dtype: int64

### `size`

In [62]:
tips["size"].dtype

dtype('int64')

In [63]:
tips["size"].value_counts()

size
2    156
3     38
4     37
5      5
1      4
6      4
Name: count, dtype: int64

In [64]:
np.sort(tips["size"].unique())

array([1, 2, 3, 4, 5, 6])

In [65]:
tips["size"].head(20).values.tolist()

[2, 3, 3, 2, 4, 4, 2, 4, 2, 2, 2, 4, 2, 4, 2, 2, 3, 3, 3, 3]

In [66]:
tips[(tips["size"] == 1) | (tips["size"] == 2) | (tips["size"] == 3) | (tips["size"] == 4) | (tips["size"] == 5) | (tips["size"] == 6)].count()

total_bill    244
tip           244
sex           244
smoker        244
day           244
time          244
size          244
dtype: int64